tests/DotNet.Testcontainers.Tests/Unit/Containers/Unix/ResourceReaperTest.cs

```diff
@@ -119,14 +124,97 @@ await this.resourceReaper.DisposeAsync()
       }
     }
 
+    [Fact]
+    public async Task ResourceReaperShouldTimeoutIfInitializationFails()
+    {
+      // Given
+      // When
+      ResourceReaper.RyukPort = 666;
+      ResourceReaper.InitTimeout = TimeSpan.FromSeconds(1);
+
+      // Then
+
+      // Do not pass timeout TimeSpan directly to CancellationTokenSource.
+      // Otherwise we can not know if the initialization can fail independently from the CancellationToken parameter.
+      using (var testTimeoutTokenSource = new CancellationTokenSource())
+      {
+        try
+        {
+          var startTask = ResourceReaper.GetAndStartNewAsync(Guid.NewGuid(), testTimeoutTokenSource.Token);
+
+          await Task.WhenAny(startTask, Task.Delay(ResourceReaper.InitTimeout.Add(TimeSpan.FromSeconds(5))));
+
+          Assert.True(startTask.IsFaulted);
+          var exception = startTask.Exception!.InnerExceptions.First();
+          Assert.IsAssignableFrom<ResourceReaperException>(exception);
+          Assert.Equal("Initialization failed", exception.Message);
+        }
+        finally
+        {
+          testTimeoutTokenSource.Cancel();
+        }
+      }
+    }
+
+    [Fact]
+    public async Task GetAndStartNewAsyncShouldBeACancellableDuringContainerStart()
+    {
+      // Given
+      // When
+      ResourceReaper.RyukPort = 666;
+      // Increase InitTimeout to ensure that GetAndStartNewAsync is stopped via the CancellationToken parameter
+      ResourceReaper.InitTimeout = TimeSpan.FromSeconds(10);
+
+      // Then
+      using (var cancellationTokenSource = new CancellationTokenSource())
+      {
+        var sessionId = Guid.NewGuid();
+        var startTask = ResourceReaper.GetAndStartNewAsync(sessionId, cancellationTokenSource.Token);
+        cancellationTokenSource.Cancel();
+
+        var exception = await Assert.ThrowsAsync<OperationCanceledException>(() => startTask);
+        Assert.Equal("The operation was canceled.", exception.Message);
+      }
+    }
+
+    [Fact]
+    public async Task GetAndStartNewAsyncShouldBeACancellableDuringConnectionInitialization()
+    {
+      // Given
+      // When
+      ResourceReaper.RyukPort = 666;
+      // Increase InitTimeout to ensure that GetAndStartNewAsync is stopped via the CancellationToken parameter
+      ResourceReaper.InitTimeout = TimeSpan.FromSeconds(10);
+
+      // Then
+      using (var cancellationTokenSource = new CancellationTokenSource())
+      {
+        var startTask = ResourceReaper.GetAndStartNewAsync(Guid.NewGuid(), cancellationTokenSource.Token);
+
+        // TODO: Is there a better way to ensure that the ryuk container is started MaintainRyukConnection is running?
+        await Task.Delay(TimeSpan.FromSeconds(2));
```

Comment:
This is quite dirty, but I had no better idea for now.

Reply:
What about an event? `ResourceReaper` can send a notification if the connection is established.

Reply:
BTW, do we need all those tests for `AutoRemove`, `CleanUp`, `CancellationToken`, etc.? I agree tests are great, but right now they are hard to understand and if the `ResourceReaper` starts properly, cleans up everything, I prefer simpler code. What do you think?

Reply:
hmm, I would keep those tests. E.g. the `AutoRemove` test failed when status `409` Http errors were not caught. In my mind it is very important to have them otherwise it will be more or less impossible to refactor/change the code in the future. I agree that they are sometimes hard to read.

The event idea is nice 👍

Reply:
Maybe we can split them later into separate classes. Then it's easier to read. But for now, let's focus on getting it stable. We can clean up the tests later.